### ISSUE TITLE
PropertyInfoLoader forces nullable=false for non built-in Doctrine Types

### ISSUE THREAD

**Issue by spackmat** [symfony/symfony#33810]
**Symfony version(s) affected**: 4.3.4

**Description**  
I am using the Automatic Validation feature and my non built-in Doctrine Type (in my case `CrEOF\Spatial\PHP\Types\Geography\Point`) always gets a `NotNull` constraint attached, ignoring that in the the column-setting, `nullable` is set to `true`.

I did some research and found out that `Symfony\Bridge\Doctrine\PropertyInfo\DoctrineExtractor::getTypes()` returns null, because it only handles built-in Doctrine Types.

After that, `Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor::getTypes()` is called and that calls `Symfony\Component\PropertyInfo\Util\PhpDocTypeHelper::getTypes()`, which defaults the `nullable` property to false and only sets it to `true`, if the non built-in Doctrine Type representation is an instance of `phpDocumentor\Reflection\Types\Nullable` or `phpDocumentor\Reflection\Types\Null_`.

In my case it is `phpDocumentor\Reflection\Types\Object_` instead so `nullable` is resolved to `false` and thus, the property with `nullable=true` of my Entity gets a `NotNull` constraint attached and **I don't know what I can do to prevent this**.

**How to reproduce**  
1. Have a Doctrine Entity with a property with a non built-in Doctrine type, which is nullable.
2. Have Automatic Validation enabled.
3. Create an instance of that Entity with the nullable property left `null`.
4. Validate that Entity and observe a validation error raised by the automatically attached `NotNull` constraint, which should not be there.

**Possible Solution**  
I am unsure if supporting custom types in `Symfony\Bridge\Doctrine\PropertyInfo\DoctrineExtractor::getTypes()` (as suggested in https://github.com/symfony/symfony/issues/29932) would solve it the "right" way, but it would prevent `Symfony\Component\PropertyInfo\PropertyInfoExtractor` from skipping it at all and let `Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor::getTypes()` set `nullable` to `false` with an incorrect assumption.
So maybe setting `nullable` to `false` by default there is the primary problem, as it leads to possibly unwanted `NotNull` constraints with Automatic Validation.

**Comment by spackmat** [comment]
Right after my post, @dunglas implemented this https://github.com/symfony/symfony/pull/32107#issuecomment-504125587 which will enable me to prevent such problems eventually. Nice, but defaulting `nullable` to `false` in `PhpDocExtractor::getTypes()` is still explosive in combination with Auto Validation. I don't have a concept of the reason for why this assumption is made and if it could be turned around or if no nullable-decision could be made at all, if the state cannot be determinded by the PhpDocExtractor. Or questioned the other way around: Is it necessary that PhpDocExtractor decides about nullable, when it has no information to rely on?

**Comment by spackmat** [comment]
To bump the issue a bit, I change my last question to a proposal:

I propose that the `PhpDocExtractor` should not decide about nullable, when it has no information to rely this decision on.

**Comment by spackmat** [comment]
I can confirm that setting `@Assert\DisableAutoMapping` on my affected property solves the symptoms. For now (Symfony 4.4.1) DisableAutoMapping fails, but with the workaround mentioned in https://github.com/symfony/symfony/issues/34672 it works.

That still doesn't solve the underlying problem as described here.

**Comment by spackmat** [comment]
With Symfony 4.4.2 released a while ago, `@Assert\DisableAutoMapping` is an acceptable workaround.

But the underlying problem is still existant: Non built-in Doctrine Types always get a NotNull constraint attached by the automapper, ignoring that in the the column-setting, nullable may be set to true.

**Comment by xabbuh** [comment]
@spackmat Would you be able to debug why that's the case and whether and how it's possible to fix it?

**Comment by spackmat** [comment]
@xabbuh I did that in the first post.

Without knowing much about the side effects and the reason, why `Symfony\Component\PropertyInfo\Util\PhpDocTypeHelper::getTypes()` defaults `nullable` to `false` (which I consider not the best decision), I would say reversing that to `true` would fix this problem (but maybe introduces others). I discussed the whole case in my posts above, so maybe the best solution would be that PhpDocExtractor doesn't decide about `nullable` at all, if it doesn't have any information to rely on. If that's possible.

**Comment by xabbuh** [comment]
Can you show how the docblock for your property looks like?

**Comment by spackmat** [comment]
Sure:

    /**
     * SRID 4326 are GPS coordinates, see http://spatialreference.org/ref/epsg/wgs-84/
     *
     * @var CrEOF\Spatial\PHP\Types\Geography\Point
     * @ORM\Column(name="location", type="geopoint", nullable=true, options={"srid"="4326"})
     * @Assert\DisableAutoMapping
     */
    private $location;

The @Assert\DisableAutoMapping is the workaround, so it must be removed to reproduce the problem.

I read into the code and the problem occurs with all non built-in Doctrine types, because `DoctrineExtractor::getTypes()` returns `null` if it steps on anything, but the build-in types. So the chain goes on an `PhpDocExtractor::getTypes()` decides to handle that property, ignores the `nullable=true` from the annotation (not even looking for it) and assumes its default of `nullable=false`, because the Point-type is not an instance of `phpDocumentor\Reflection\Types\Nullable` or `phpDocumentor\Reflection\Types\Null_` (what is its only considered input about `nullable`).

**Comment by xabbuh** [comment]
```php
/**
 * @var CrEOF\Spatial\PHP\Types\Geography\Point|null
 */
```

Does that change make it work (you can keep the other annotations besides `DisableAutoMapping`)?

**Comment by spackmat** [comment]
Yes that solves the problem, removing the `@var` also works. Thank you!

Wow, didn't see that all the time (and misunderstood the for loop in `PhpDocTypeHelper::getTypes()`)! None of my other nullable properties are documented that way. To sum that up: The `DoctrineExtractor` ignores non-built-in types, so the `@ORM\Column` is not considered for property extraction on that cases. Instead the `PhpDocExtractor` kicks in, sees the `@var` without `null` mentioned and deduces `nullable=false` from that. What ist correct from its point of view, but in spite inconsistent with the column definition.

So the problem reduces to the fact that the `DoctrineExtractor` doesn't feel that need to extract options from non-built-in types and skips them, so an incorrect `@var` annotation wins the inconsistency-dispute over the `nullable=true` from the column definition.

So could it be an option to change the `DoctrineExtrator` to not skip all non-built-in types and read their explicitly set options? Since auto-validation-mapping should use column definitions first.

Or just close this issue as this only occurs if some stupid devs produce such inconsistent definitions.

**Comment by xabbuh** [comment]
> So could it be an option to change the `DoctrineExtrator` to not skip all non-built-in types and read their explicitly set options? Since auto-validation-mapping should use column definitions first.

We should investigate if we are able to read the information for custom types and if so update the code accordingly.

That does not solve issues when information provided by different sources is inconsistent, but may mitigate some issues.

**Comment by stof** [comment]
The thing is, for non-builtin types, the extractor won't be able to know what PHP type will be used for the property (how can it know that your custom type turns things into a Point object, while it cannot know your custom type as it is custom ?)
There is nothing in the Doctrine mapping referencing the fact that your custom `geopoint` type produces a Point object (because Doctrine does not care about that info, it passes the value to your type for conversion).

**Comment by spackmat** [comment]
I see, then I close the issue and we hope that some other stupid devs with similar inconsistent annotations will find it in the future. :) Thanks for clarification!

**Comment by spackmat** [comment]
Hello again. I just encountered the old problem with a (builtin) `json` doctrine type field holding a collection: Without `@Assert\DisableAutoMapping` I got an unwanted `NotNull` constraint attached to my field that has `nullable=true`.

It turned out that I (again) had a `@var array` annotation on that property and thus the extractor assumed it as not nullable. Changing that to `@var array|null` solved this and now I hope my repeated stupidity helps others to not spend several hours in tears. ;)
